{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/riuc13/Proyecto-final/blob/main/Copia_de_AlphaFold.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "pc5-mbsX9PZC"
      },
      "source": [
        "# AlphaFold Colab\n",
        "\n",
        "Este cuaderno de Colab te permite predecir fácilmente la estructura de una proteína utilizando una versión ligeramente simplificada de [AlphaFold v2.3.1](https://doi.org/10.1038/s41586-021-03819-2). \n",
        "\n",
        "**Diferencias con AlphaFold v2.3.1**\n",
        "\n",
        "En comparación con AlphaFold v2.3.1, este cuaderno de Colab no utiliza plantillas (estructuras homólogas) y una parte seleccionada de la [BFD database](https://bfd.mmseqs.com/). Hemos validado estos cambios en varios miles de estructuras PDB recientes. Aunque la precisión será prácticamente idéntica al sistema completo de AlphaFold en muchos objetivos, una pequeña fracción presenta una disminución importante en la precisión debido a la MSA más pequeña y la falta de plantillas. Para mayor fiabilidad, recomendamos utilizar en [full open source AlphaFold](https://github.com/deepmind/alphafold/), o la [AlphaFold Protein Structure Database](https://alphafold.ebi.ac.uk/).\n",
        "\n",
        "**Este Colab tiene una pequeña disminución en la precisión promedio para multimers en comparación con la instalación local de AlphaFold. Para una precisión completa de multímeros, se recomienda encarecidamente ejecutar  [AlphaFold locally](https://github.com/deepmind/alphafold#running-alphafold).** Además, AlphaFold-Multímero requiere buscar MSA para cada secuencia única en el complejo, por lo que es sustancialmente más lento. Si su cuaderno se agota por la búsqueda lenta de MSA de multímeros, recomendamos utilizar Colab Pro o ejecutar AlphaFold localmente.\n",
        "\n",
        "Tenga en cuenta que este cuaderno de Colab se proporciona solo para modelado teórico y se deben tener precaución en su uso. \n",
        "\n",
        "El **formato de archivo PAE** se ha actualizado para coincidir con AFDB. Consulte la [AFDB FAQ](https://alphafold.ebi.ac.uk/faq/#faq-7) para obtener una descripción del nuevo formato.\n",
        "\n",
        "**Citar esta obra**\n",
        "\n",
        "Cualquier publicación que revele hallazgos derivados del uso de este cuaderno debería [cite](https://github.com/deepmind/alphafold/#citing-this-work) el [AlphaFold paper](https://doi.org/10.1038/s41586-021-03819-2).\n",
        "\n",
        "**Licencias**\n",
        "\n",
        "Este Colab utiliza los parámetros del [AlphaFold model parameters](https://github.com/deepmind/alphafold/#model-parameters-license) que están sujetos a la licencia Creative Commons Attribution 4.0 International ([CC BY 4.0](https://creativecommons.org/licenses/by/4.0/legalcode)). El Colab en sí se proporciona bajo la  [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0). Vea la declaración completa de licencia a continuación.\n",
        "\n",
        "**Más información**\n",
        "\n",
        "Puedes encontrar más información sobre cómo funciona AlphaFold en los siguientes artículos:\n",
        "\n",
        "*   [AlphaFold methods paper](https://www.nature.com/articles/s41586-021-03819-2)\n",
        "*   [AlphaFold predictions of the human proteome paper](https://www.nature.com/articles/s41586-021-03828-1)\n",
        "*   [AlphaFold-Multimer paper](https://www.biorxiv.org/content/10.1101/2021.10.04.463034v1)\n",
        "\n",
        "Las preguntas frecuentes sobre cómo interpretar las predicciones de AlphaFold son [here](https://alphafold.ebi.ac.uk/faq).\n",
        "\n",
        "Si tiene alguna pregunta que no esté cubierta en las preguntas frecuentes, comuníquese con el equipo de AlphaFold en alphafold@deepmind.com.\n",
        "\n",
        "**Ponerse en contacto**\n",
        "\n",
        "Nos encantaría escuchar sus comentarios y entender cómo AlphaFold ha sido útil en su investigación. Comparte tus historias con nosotros en alphafold@deepmind.com.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "uC1dKAwk2eyl"
      },
      "source": [
        "## configuración\n",
        "\n",
        "Comience ejecutando las 2 celdas a continuación para configurar AlphaFold y todo el software necesario."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "woIxeCPygt7K"
      },
      "outputs": [],
      "source": [
        "# Set environment variables before running any other code.\n",
        "import os\n",
        "os.environ['TF_FORCE_UNIFIED_MEMORY'] = '1'\n",
        "os.environ['XLA_PYTHON_CLIENT_MEM_FRACTION'] = '4.0'\n",
        "\n",
        "#@title 1. Instalar software de terceros\n",
        "\n",
        "#@markdown Ejecute esta celda presionando el botón Reproducir a la izquierda para descargar e importar software de terceros en este cuaderno de Colab.(Consulte los [acknowledgements](https://github.com/deepmind/alphafold/#acknowledgements) en nuestro archivo Léame).\n",
        "\n",
        "#@markdown **Nota**: Esto instala el software en la computadora portátil Colab en la nube y no en su computadora.\n",
        "\n",
        "from IPython.utils import io\n",
        "import os\n",
        "import subprocess\n",
        "import tqdm.notebook\n",
        "\n",
        "TQDM_BAR_FORMAT = '{l_bar}{bar}| {n_fmt}/{total_fmt} [elapsed: {elapsed} remaining: {remaining}]'\n",
        "\n",
        "try:\n",
        "  with tqdm.notebook.tqdm(total=100, bar_format=TQDM_BAR_FORMAT) as pbar:\n",
        "    with io.capture_output() as captured:\n",
        "      # Uninstall default Colab version of TF.\n",
        "      %shell pip uninstall -y tensorflow\n",
        "\n",
        "      %shell sudo apt install --quiet --yes hmmer\n",
        "      pbar.update(6)\n",
        "\n",
        "      # Install py3dmol.\n",
        "      %shell pip install py3dmol\n",
        "      pbar.update(2)\n",
        "\n",
        "      # Install OpenMM and pdbfixer.\n",
        "      %shell rm -rf /opt/conda\n",
        "      %shell wget -q -P /tmp \\\n",
        "        https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh \\\n",
        "          && bash /tmp/Miniconda3-latest-Linux-x86_64.sh -b -p /opt/conda \\\n",
        "          && rm /tmp/Miniconda3-latest-Linux-x86_64.sh\n",
        "      pbar.update(9)\n",
        "\n",
        "      PATH=%env PATH\n",
        "      %env PATH=/opt/conda/bin:{PATH}\n",
        "      %shell conda install -qy conda==4.13.0 \\\n",
        "          && conda install -qy -c conda-forge \\\n",
        "            python=3.8 \\\n",
        "            openmm=7.5.1 \\\n",
        "            pdbfixer\n",
        "      pbar.update(80)\n",
        "\n",
        "      # Create a ramdisk to store a database chunk to make Jackhmmer run fast.\n",
        "      %shell sudo mkdir -m 777 --parents /tmp/ramdisk\n",
        "      %shell sudo mount -t tmpfs -o size=9G ramdisk /tmp/ramdisk\n",
        "      pbar.update(2)\n",
        "\n",
        "      %shell wget -q -P /content \\\n",
        "        https://git.scicore.unibas.ch/schwede/openstructure/-/raw/7102c63615b64735c4941278d92b554ec94415f8/modules/mol/alg/src/stereo_chemical_props.txt\n",
        "      pbar.update(1)\n",
        "except subprocess.CalledProcessError:\n",
        "  print(captured)\n",
        "  raise"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "VzJ5iMjTtoZw"
      },
      "outputs": [],
      "source": [
        "#@title 2. Descargar AlphaFold\n",
        "\n",
        "#@markdown Ejecute esta celda presionando el botón *Reproducir* en\n",
        "#@markdown la izquierda.\n",
        "\n",
        "GIT_REPO = 'https://github.com/deepmind/alphafold'\n",
        "SOURCE_URL = 'https://storage.googleapis.com/alphafold/alphafold_params_colab_2022-12-06.tar'\n",
        "PARAMS_DIR = './alphafold/data/params'\n",
        "PARAMS_PATH = os.path.join(PARAMS_DIR, os.path.basename(SOURCE_URL))\n",
        "\n",
        "try:\n",
        "  with tqdm.notebook.tqdm(total=100, bar_format=TQDM_BAR_FORMAT) as pbar:\n",
        "    with io.capture_output() as captured:\n",
        "      %shell rm -rf alphafold\n",
        "      %shell git clone --branch main {GIT_REPO} alphafold\n",
        "      pbar.update(8)\n",
        "      # Install the required versions of all dependencies.\n",
        "      %shell pip3 install -r ./alphafold/requirements.txt\n",
        "      # Run setup.py to install only AlphaFold.\n",
        "      %shell pip3 install --no-dependencies ./alphafold\n",
        "      pbar.update(10)\n",
        "\n",
        "      # Apply OpenMM patch.\n",
        "      %shell pushd /opt/conda/lib/python3.8/site-packages/ && \\\n",
        "          patch -p0 < /content/alphafold/docker/openmm.patch && \\\n",
        "          popd\n",
        "\n",
        "      # Make sure stereo_chemical_props.txt is in all locations where it could be searched for.\n",
        "      %shell mkdir -p /content/alphafold/alphafold/common\n",
        "      %shell cp -f /content/stereo_chemical_props.txt /content/alphafold/alphafold/common\n",
        "      %shell mkdir -p /opt/conda/lib/python3.8/site-packages/alphafold/common/\n",
        "      %shell cp -f /content/stereo_chemical_props.txt /opt/conda/lib/python3.8/site-packages/alphafold/common/\n",
        "\n",
        "      # Load parameters\n",
        "      %shell mkdir --parents \"{PARAMS_DIR}\"\n",
        "      %shell wget -O \"{PARAMS_PATH}\" \"{SOURCE_URL}\"\n",
        "      pbar.update(27)\n",
        "\n",
        "      %shell tar --extract --verbose --file=\"{PARAMS_PATH}\" \\\n",
        "        --directory=\"{PARAMS_DIR}\" --preserve-permissions\n",
        "      %shell rm \"{PARAMS_PATH}\"\n",
        "      pbar.update(55)\n",
        "except subprocess.CalledProcessError:\n",
        "  print(captured)\n",
        "  raise\n",
        "\n",
        "import jax\n",
        "if jax.local_devices()[0].platform == 'tpu':\n",
        "  raise RuntimeError('Colab TPU runtime not supported. Change it to GPU via Runtime -> Change Runtime Type -> Hardware accelerator -> GPU.')\n",
        "elif jax.local_devices()[0].platform == 'cpu':\n",
        "  raise RuntimeError('Colab CPU runtime not supported. Change it to GPU via Runtime -> Change Runtime Type -> Hardware accelerator -> GPU.')\n",
        "else:\n",
        "  print(f'Running with {jax.local_devices()[0].device_kind} GPU')\n",
        "\n",
        "# Make sure everything we need is on the path.\n",
        "import sys\n",
        "sys.path.append('/opt/conda/lib/python3.8/site-packages')\n",
        "sys.path.append('/content/alphafold')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "W4JpOs6oA-QS"
      },
      "source": [
        "## Haciendo una predicción\n",
        "\n",
        "Pegue la secuencia de su proteína en el cuadro de texto a continuación, luego ejecute las celdas restantes a través de _Runtime_ > _Run after_. También puede ejecutar las celdas individualmente presionando el botón _Reproducir_ a la izquierda.\n",
        "\n",
        "Tenga en cuenta que la búsqueda en las bases de datos y la predicción real pueden llevar algún tiempo, de minutos a horas, según la longitud de la proteína y el tipo de GPU que Colab le asigne (consulte las preguntas frecuentes a continuación)."
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "rowN0bVYLe9n"
      },
      "outputs": [],
      "source": [
        "#@title 3. Introduzca la(s) secuencia(s) de aminoácidos a plegar ⬇️\n",
        "#@markdown Ingrese la(s) secuencia(s) de aminoácidos a plegar:\n",
        "#@markdown * Si ingresa solo una secuencia, el modelo de monómero será\n",
        "#@markdown utilizado (a menos que anule esto a continuación).\n",
        "#@markdown * Si ingresa varias secuencias, se usará el modelo multimérico.\n",
        "\n",
        "from alphafold.notebooks import notebook_utils\n",
        "import enum\n",
        "\n",
        "@enum.unique\n",
        "class ModelType(enum.Enum):\n",
        "  MONOMER = 0\n",
        "  MULTIMER = 1\n",
        "\n",
        "sequence_1 = 'MAAHKGAEHHHKAAEHHEQAAKHHHAAAEHHEKGEHEQAAHHADTAYAHHKHAEEHAAQAAKHDAEHHAPKPH'  #@param {type:\"string\"}\n",
        "sequence_2 = ''  #@param {type:\"string\"}\n",
        "sequence_3 = ''  #@param {type:\"string\"}\n",
        "sequence_4 = ''  #@param {type:\"string\"}\n",
        "sequence_5 = ''  #@param {type:\"string\"}\n",
        "sequence_6 = ''  #@param {type:\"string\"}\n",
        "sequence_7 = ''  #@param {type:\"string\"}\n",
        "sequence_8 = ''  #@param {type:\"string\"}\n",
        "sequence_9 = ''  #@param {type:\"string\"}\n",
        "sequence_10 = ''  #@param {type:\"string\"}\n",
        "sequence_11 = ''  #@param {type:\"string\"}\n",
        "sequence_12 = ''  #@param {type:\"string\"}\n",
        "sequence_13 = ''  #@param {type:\"string\"}\n",
        "sequence_14 = ''  #@param {type:\"string\"}\n",
        "sequence_15 = ''  #@param {type:\"string\"}\n",
        "sequence_16 = ''  #@param {type:\"string\"}\n",
        "sequence_17 = ''  #@param {type:\"string\"}\n",
        "sequence_18 = ''  #@param {type:\"string\"}\n",
        "sequence_19 = ''  #@param {type:\"string\"}\n",
        "sequence_20 = ''  #@param {type:\"string\"}\n",
        "\n",
        "input_sequences = (\n",
        "    sequence_1, sequence_2, sequence_3, sequence_4, sequence_5, \n",
        "    sequence_6, sequence_7, sequence_8, sequence_9, sequence_10,\n",
        "    sequence_11, sequence_12, sequence_13, sequence_14, sequence_15, \n",
        "    sequence_16, sequence_17, sequence_18, sequence_19, sequence_20)\n",
        "\n",
        "MIN_PER_SEQUENCE_LENGTH = 16\n",
        "MAX_PER_SEQUENCE_LENGTH = 4000\n",
        "MAX_MONOMER_MODEL_LENGTH = 2500\n",
        "MAX_LENGTH = 4000\n",
        "MAX_VALIDATED_LENGTH = 3000\n",
        "\n",
        "#@markdown Seleccione esta casilla de verificación para ejecutar el modelo multimérico para una sola secuencia.\n",
        "#@markdown Para proteínas que son monoméricas en su forma nativa, o para proteínas muy\n",
        "#@markdown grandes cadenas individuales puede obtener una mayor precisión y eficiencia de la memoria\n",
        "#@markdown utilizando el modelo multimérico.\n",
        "#@markdown \n",
        "#@markdown \n",
        "#@markdown Debido a la mejora de la eficiencia de la memoria, el modelo multímero tiene un máximo\n",
        "#@markdown límite de 4000 residuos, mientras que el modelo de monómero tiene un límite de 2500\n",
        "#@markdown residuos\n",
        "\n",
        "use_multimer_model_for_monomers = False #@param {type:\"boolean\"}\n",
        "\n",
        "# Validate the input sequences.\n",
        "sequences = notebook_utils.clean_and_validate_input_sequences(\n",
        "    input_sequences=input_sequences,\n",
        "    min_sequence_length=MIN_PER_SEQUENCE_LENGTH,\n",
        "    max_sequence_length=MAX_PER_SEQUENCE_LENGTH)\n",
        "\n",
        "if len(sequences) == 1:\n",
        "  if use_multimer_model_for_monomers:\n",
        "    print('Using the multimer model for single-chain, as requested.')\n",
        "    model_type_to_use = ModelType.MULTIMER\n",
        "  else:\n",
        "    print('Using the single-chain model.')\n",
        "    model_type_to_use = ModelType.MONOMER\n",
        "else:\n",
        "  print(f'Using the multimer model with {len(sequences)} sequences.')\n",
        "  model_type_to_use = ModelType.MULTIMER\n",
        "\n",
        "# Check whether total length exceeds limit.\n",
        "total_sequence_length = sum([len(seq) for seq in sequences])\n",
        "if total_sequence_length > MAX_LENGTH:\n",
        "  raise ValueError('The total sequence length is too long: '\n",
        "                   f'{total_sequence_length}, while the maximum is '\n",
        "                   f'{MAX_LENGTH}.')\n",
        "\n",
        "# Check whether we exceed the monomer limit.\n",
        "if model_type_to_use == ModelType.MONOMER:\n",
        "  if len(sequences[0]) > MAX_MONOMER_MODEL_LENGTH:\n",
        "    raise ValueError(\n",
        "        f'Input sequence is too long: {len(sequences[0])} amino acids, while '\n",
        "        f'the maximum for the monomer model is {MAX_MONOMER_MODEL_LENGTH}. You may '\n",
        "        'be able to run this sequence with the multimer model by selecting the '\n",
        "        'use_multimer_model_for_monomers checkbox above.')\n",
        "    \n",
        "if total_sequence_length > MAX_VALIDATED_LENGTH:\n",
        "  print('WARNING: The accuracy of the system has not been fully validated '\n",
        "        'above 3000 residues, and you may experience long running times or '\n",
        "        f'run out of memory. Total sequence length is {total_sequence_length} '\n",
        "        'residues.')\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "2tTeTTsLKPjB"
      },
      "outputs": [],
      "source": [
        "#@title 4. Búsqueda en bases de datos genéticas\n",
        "\n",
        "#@markdown Una vez que se haya ejecutado esta celda, verá\n",
        "#@markdown estadísticas sobre la alineación de secuencias múltiples\n",
        "#@markdown (MSA) que utilizará AlphaFold. En particular,\n",
        "#@markdown verá qué tan bien cada residuo está cubierto por similares\n",
        "#@markdown secuencias en el MSA.\n",
        "\n",
        "# --- Python imports ---\n",
        "import collections\n",
        "import copy\n",
        "from concurrent import futures\n",
        "import json\n",
        "import random\n",
        "import shutil\n",
        "\n",
        "from urllib import request\n",
        "from google.colab import files\n",
        "from matplotlib import gridspec\n",
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "import py3Dmol\n",
        "\n",
        "from alphafold.model import model\n",
        "from alphafold.model import config\n",
        "from alphafold.model import data\n",
        "\n",
        "from alphafold.data import feature_processing\n",
        "from alphafold.data import msa_pairing\n",
        "from alphafold.data import pipeline\n",
        "from alphafold.data import pipeline_multimer\n",
        "from alphafold.data.tools import jackhmmer\n",
        "\n",
        "from alphafold.common import protein\n",
        "\n",
        "from alphafold.relax import relax\n",
        "from alphafold.relax import utils\n",
        "\n",
        "from IPython import display\n",
        "from ipywidgets import GridspecLayout\n",
        "from ipywidgets import Output\n",
        "\n",
        "# Color bands for visualizing plddt\n",
        "PLDDT_BANDS = [(0, 50, '#FF7D45'),\n",
        "               (50, 70, '#FFDB13'),\n",
        "               (70, 90, '#65CBF3'),\n",
        "               (90, 100, '#0053D6')]\n",
        "\n",
        "# --- Find the closest source ---\n",
        "test_url_pattern = 'https://storage.googleapis.com/alphafold-colab{:s}/latest/uniref90_2022_01.fasta.1'\n",
        "ex = futures.ThreadPoolExecutor(3)\n",
        "def fetch(source):\n",
        "  request.urlretrieve(test_url_pattern.format(source))\n",
        "  return source\n",
        "fs = [ex.submit(fetch, source) for source in ['', '-europe', '-asia']]\n",
        "source = None\n",
        "for f in futures.as_completed(fs):\n",
        "  source = f.result()\n",
        "  ex.shutdown()\n",
        "  break\n",
        "\n",
        "JACKHMMER_BINARY_PATH = '/usr/bin/jackhmmer'\n",
        "DB_ROOT_PATH = f'https://storage.googleapis.com/alphafold-colab{source}/latest/'\n",
        "# The z_value is the number of sequences in a database.\n",
        "MSA_DATABASES = [\n",
        "    {'db_name': 'uniref90',\n",
        "     'db_path': f'{DB_ROOT_PATH}uniref90_2022_01.fasta',\n",
        "     'num_streamed_chunks': 62,\n",
        "     'z_value': 144_113_457},\n",
        "    {'db_name': 'smallbfd',\n",
        "     'db_path': f'{DB_ROOT_PATH}bfd-first_non_consensus_sequences.fasta',\n",
        "     'num_streamed_chunks': 17,\n",
        "     'z_value': 65_984_053},\n",
        "    {'db_name': 'mgnify',\n",
        "     'db_path': f'{DB_ROOT_PATH}mgy_clusters_2022_05.fasta',\n",
        "     'num_streamed_chunks': 120,\n",
        "     'z_value': 623_796_864},\n",
        "]\n",
        "\n",
        "# Search UniProt and construct the all_seq features only for heteromers, not homomers.\n",
        "if model_type_to_use == ModelType.MULTIMER and len(set(sequences)) > 1:\n",
        "  MSA_DATABASES.extend([\n",
        "      # Swiss-Prot and TrEMBL are concatenated together as UniProt.\n",
        "      {'db_name': 'uniprot',\n",
        "       'db_path': f'{DB_ROOT_PATH}uniprot_2021_04.fasta',\n",
        "       'num_streamed_chunks': 101,\n",
        "       'z_value': 225_013_025 + 565_928},\n",
        "  ])\n",
        "\n",
        "TOTAL_JACKHMMER_CHUNKS = sum([cfg['num_streamed_chunks'] for cfg in MSA_DATABASES])\n",
        "\n",
        "MAX_HITS = {\n",
        "    'uniref90': 10_000,\n",
        "    'smallbfd': 5_000,\n",
        "    'mgnify': 501,\n",
        "    'uniprot': 50_000,\n",
        "}\n",
        "\n",
        "\n",
        "def get_msa(sequences):\n",
        "  \"\"\"Searches for MSA for given sequences using chunked Jackhmmer search.\n",
        "  \n",
        "  Args:\n",
        "    sequences: A list of sequences to search against all databases.\n",
        "\n",
        "  Returns:\n",
        "    A dictionary mapping unique sequences to dicionaries mapping each database\n",
        "    to a list of  results, one for each chunk of the database.\n",
        "  \"\"\"\n",
        "  sequence_to_fasta_path = {}\n",
        "  # Deduplicate to not do redundant work for multiple copies of the same chain in homomers.\n",
        "  for sequence_index, sequence in enumerate(sorted(set(sequences)), 1):\n",
        "    fasta_path = f'target_{sequence_index:02d}.fasta'\n",
        "    with open(fasta_path, 'wt') as f:\n",
        "      f.write(f'>query\\n{sequence}')\n",
        "    sequence_to_fasta_path[sequence] = fasta_path\n",
        "\n",
        "  # Run the search against chunks of genetic databases (since the genetic\n",
        "  # databases don't fit in Colab disk).\n",
        "  raw_msa_results = {sequence: {} for sequence in sequence_to_fasta_path.keys()}\n",
        "  print('\\nGetting MSA for all sequences')\n",
        "  with tqdm.notebook.tqdm(total=TOTAL_JACKHMMER_CHUNKS, bar_format=TQDM_BAR_FORMAT) as pbar:\n",
        "    def jackhmmer_chunk_callback(i):\n",
        "      pbar.update(n=1)\n",
        "\n",
        "    for db_config in MSA_DATABASES:\n",
        "      db_name = db_config['db_name']\n",
        "      pbar.set_description(f'Searching {db_name}')\n",
        "      jackhmmer_runner = jackhmmer.Jackhmmer(\n",
        "          binary_path=JACKHMMER_BINARY_PATH,\n",
        "          database_path=db_config['db_path'],\n",
        "          get_tblout=True,\n",
        "          num_streamed_chunks=db_config['num_streamed_chunks'],\n",
        "          streaming_callback=jackhmmer_chunk_callback,\n",
        "          z_value=db_config['z_value'])\n",
        "      # Query all unique sequences against each chunk of the database to prevent\n",
        "      # redunantly fetching each chunk for each unique sequence.\n",
        "      results = jackhmmer_runner.query_multiple(list(sequence_to_fasta_path.values()))\n",
        "      for sequence, result_for_sequence in zip(sequence_to_fasta_path.keys(), results):\n",
        "        raw_msa_results[sequence][db_name] = result_for_sequence\n",
        "\n",
        "  return raw_msa_results\n",
        "\n",
        "\n",
        "features_for_chain = {}\n",
        "raw_msa_results_for_sequence = get_msa(sequences)\n",
        "for sequence_index, sequence in enumerate(sequences, start=1):\n",
        "  raw_msa_results = copy.deepcopy(raw_msa_results_for_sequence[sequence])\n",
        "\n",
        "  # Extract the MSAs from the Stockholm files.\n",
        "  # NB: deduplication happens later in pipeline.make_msa_features.\n",
        "  single_chain_msas = []\n",
        "  uniprot_msa = None\n",
        "  for db_name, db_results in raw_msa_results.items():\n",
        "    merged_msa = notebook_utils.merge_chunked_msa(\n",
        "        results=db_results, max_hits=MAX_HITS.get(db_name))\n",
        "    if merged_msa.sequences and db_name != 'uniprot':\n",
        "      single_chain_msas.append(merged_msa)\n",
        "      msa_size = len(set(merged_msa.sequences))\n",
        "      print(f'{msa_size} unique sequences found in {db_name} for sequence {sequence_index}')\n",
        "    elif merged_msa.sequences and db_name == 'uniprot':\n",
        "      uniprot_msa = merged_msa\n",
        "\n",
        "  notebook_utils.show_msa_info(single_chain_msas=single_chain_msas, sequence_index=sequence_index)\n",
        "\n",
        "  # Turn the raw data into model features.\n",
        "  feature_dict = {}\n",
        "  feature_dict.update(pipeline.make_sequence_features(\n",
        "      sequence=sequence, description='query', num_res=len(sequence)))\n",
        "  feature_dict.update(pipeline.make_msa_features(msas=single_chain_msas))\n",
        "  # We don't use templates in AlphaFold Colab notebook, add only empty placeholder features.\n",
        "  feature_dict.update(notebook_utils.empty_placeholder_template_features(\n",
        "      num_templates=0, num_res=len(sequence)))\n",
        "\n",
        "  # Construct the all_seq features only for heteromers, not homomers.\n",
        "  if model_type_to_use == ModelType.MULTIMER and len(set(sequences)) > 1:\n",
        "    valid_feats = msa_pairing.MSA_FEATURES + (\n",
        "        'msa_species_identifiers',\n",
        "    )\n",
        "    all_seq_features = {\n",
        "        f'{k}_all_seq': v for k, v in pipeline.make_msa_features([uniprot_msa]).items()\n",
        "        if k in valid_feats}\n",
        "    feature_dict.update(all_seq_features)\n",
        "\n",
        "  features_for_chain[protein.PDB_CHAIN_IDS[sequence_index - 1]] = feature_dict\n",
        "\n",
        "\n",
        "# Do further feature post-processing depending on the model type.\n",
        "if model_type_to_use == ModelType.MONOMER:\n",
        "  np_example = features_for_chain[protein.PDB_CHAIN_IDS[0]]\n",
        "\n",
        "elif model_type_to_use == ModelType.MULTIMER:\n",
        "  all_chain_features = {}\n",
        "  for chain_id, chain_features in features_for_chain.items():\n",
        "    all_chain_features[chain_id] = pipeline_multimer.convert_monomer_features(\n",
        "        chain_features, chain_id)\n",
        "\n",
        "  all_chain_features = pipeline_multimer.add_assembly_features(all_chain_features)\n",
        "\n",
        "  np_example = feature_processing.pair_and_merge(\n",
        "      all_chain_features=all_chain_features)\n",
        "\n",
        "  # Pad MSA to avoid zero-sized extra_msa.\n",
        "  np_example = pipeline_multimer.pad_msa(np_example, min_num_seq=512)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "cellView": "form",
        "id": "XUo6foMQxwS2"
      },
      "outputs": [],
      "source": [
        "#@title 5. Ejecute AlphaFold y descargue la predicción\n",
        "\n",
        "#@markdown Una vez que se ha ejecutado esta celda, un archivo zip con\n",
        "#@markdown la predicción obtenida se descargará automáticamente\n",
        "#@markdown a tu computador.\n",
        "\n",
        "#@markdown En caso de que tenga problemas con la etapa de relajación, puede desactivarla a continuación.\n",
        "#@markdown Advertencia: Esto significa que la predicción puede tener efectos de distracción.\n",
        "#@markdown pequeñas violaciones estereoquímicas.\n",
        "\n",
        "run_relax = False  #@param {type:\"boolean\"}\n",
        "\n",
        "#@markdown La relajación es más rápida con una GPU, pero hemos descubierto que es menos estable.\n",
        "#@markdown Es posible que desee habilitar la GPU para obtener un mayor rendimiento, pero si no lo hace\n",
        "#@markdown converge sugerimos volver a usar sin GPU.\n",
        "\n",
        "relax_use_gpu = False  #@param {type:\"boolean\"}\n",
        "\n",
        "\n",
        "#@markdown El modelo de multímeros seguirá reciclando hasta que se detengan las predicciones\n",
        "#@markdown cambiando, hasta el límite establecido aquí. Para una mayor precisión, en el\n",
        "#@markdown costo potencial de tiempos de inferencia más largos, establezca esto en 20.\n",
        "\n",
        "multimer_model_max_num_recycles = 3  #@param {type:\"integer\"}\n",
        "\n",
        "\n",
        "# --- Run the model ---\n",
        "if model_type_to_use == ModelType.MONOMER:\n",
        "  model_names = config.MODEL_PRESETS['monomer'] + ('model_2_ptm',)\n",
        "elif model_type_to_use == ModelType.MULTIMER:\n",
        "  model_names = config.MODEL_PRESETS['multimer']\n",
        "\n",
        "output_dir = 'prediction'\n",
        "os.makedirs(output_dir, exist_ok=True)\n",
        "\n",
        "plddts = {}\n",
        "ranking_confidences = {}\n",
        "pae_outputs = {}\n",
        "unrelaxed_proteins = {}\n",
        "\n",
        "with tqdm.notebook.tqdm(total=len(model_names) + 1, bar_format=TQDM_BAR_FORMAT) as pbar:\n",
        "  for model_name in model_names:\n",
        "    pbar.set_description(f'Running {model_name}')\n",
        "\n",
        "    cfg = config.model_config(model_name)\n",
        "\n",
        "    if model_type_to_use == ModelType.MONOMER:\n",
        "      cfg.data.eval.num_ensemble = 1\n",
        "    elif model_type_to_use == ModelType.MULTIMER:\n",
        "      cfg.model.num_ensemble_eval = 1\n",
        "\n",
        "    if model_type_to_use == ModelType.MULTIMER:\n",
        "      cfg.model.num_recycle = multimer_model_max_num_recycles\n",
        "      cfg.model.recycle_early_stop_tolerance = 0.5\n",
        "\n",
        "    params = data.get_model_haiku_params(model_name, './alphafold/data')\n",
        "    model_runner = model.RunModel(cfg, params)\n",
        "    processed_feature_dict = model_runner.process_features(np_example, random_seed=0)\n",
        "    prediction = model_runner.predict(processed_feature_dict, random_seed=random.randrange(sys.maxsize))\n",
        "\n",
        "    mean_plddt = prediction['plddt'].mean()\n",
        "\n",
        "    if model_type_to_use == ModelType.MONOMER:\n",
        "      if 'predicted_aligned_error' in prediction:\n",
        "        pae_outputs[model_name] = (prediction['predicted_aligned_error'],\n",
        "                                   prediction['max_predicted_aligned_error'])\n",
        "      else:\n",
        "        # Monomer models are sorted by mean pLDDT. Do not put monomer pTM models here as they\n",
        "        # should never get selected.\n",
        "        ranking_confidences[model_name] = prediction['ranking_confidence']\n",
        "        plddts[model_name] = prediction['plddt']\n",
        "    elif model_type_to_use == ModelType.MULTIMER:\n",
        "      # Multimer models are sorted by pTM+ipTM.\n",
        "      ranking_confidences[model_name] = prediction['ranking_confidence']\n",
        "      plddts[model_name] = prediction['plddt']\n",
        "      pae_outputs[model_name] = (prediction['predicted_aligned_error'],\n",
        "                                 prediction['max_predicted_aligned_error'])\n",
        "\n",
        "    # Set the b-factors to the per-residue plddt.\n",
        "    final_atom_mask = prediction['structure_module']['final_atom_mask']\n",
        "    b_factors = prediction['plddt'][:, None] * final_atom_mask\n",
        "    unrelaxed_protein = protein.from_prediction(\n",
        "        processed_feature_dict,\n",
        "        prediction,\n",
        "        b_factors=b_factors,\n",
        "        remove_leading_feature_dimension=(\n",
        "            model_type_to_use == ModelType.MONOMER))\n",
        "    unrelaxed_proteins[model_name] = unrelaxed_protein\n",
        "\n",
        "    # Delete unused outputs to save memory.\n",
        "    del model_runner\n",
        "    del params\n",
        "    del prediction\n",
        "    pbar.update(n=1)\n",
        "\n",
        "  # --- AMBER relax the best model ---\n",
        "\n",
        "  # Find the best model according to the mean pLDDT.\n",
        "  best_model_name = max(ranking_confidences.keys(), key=lambda x: ranking_confidences[x])\n",
        "\n",
        "  if run_relax:\n",
        "    pbar.set_description(f'AMBER relaxation')\n",
        "    amber_relaxer = relax.AmberRelaxation(\n",
        "        max_iterations=0,\n",
        "        tolerance=2.39,\n",
        "        stiffness=10.0,\n",
        "        exclude_residues=[],\n",
        "        max_outer_iterations=3,\n",
        "        use_gpu=relax_use_gpu)\n",
        "    relaxed_pdb, _, _ = amber_relaxer.process(prot=unrelaxed_proteins[best_model_name])\n",
        "  else:\n",
        "    print('Warning: Running without the relaxation stage.')\n",
        "    relaxed_pdb = protein.to_pdb(unrelaxed_proteins[best_model_name])\n",
        "  pbar.update(n=1)  # Finished AMBER relax.\n",
        "\n",
        "# Construct multiclass b-factors to indicate confidence bands\n",
        "# 0=very low, 1=low, 2=confident, 3=very high\n",
        "banded_b_factors = []\n",
        "for plddt in plddts[best_model_name]:\n",
        "  for idx, (min_val, max_val, _) in enumerate(PLDDT_BANDS):\n",
        "    if plddt >= min_val and plddt <= max_val:\n",
        "      banded_b_factors.append(idx)\n",
        "      break\n",
        "banded_b_factors = np.array(banded_b_factors)[:, None] * final_atom_mask\n",
        "to_visualize_pdb = utils.overwrite_b_factors(relaxed_pdb, banded_b_factors)\n",
        "\n",
        "\n",
        "# Write out the prediction\n",
        "pred_output_path = os.path.join(output_dir, 'selected_prediction.pdb')\n",
        "with open(pred_output_path, 'w') as f:\n",
        "  f.write(relaxed_pdb)\n",
        "\n",
        "\n",
        "# --- Visualise the prediction & confidence ---\n",
        "show_sidechains = True\n",
        "def plot_plddt_legend():\n",
        "  \"\"\"Plots the legend for pLDDT.\"\"\"\n",
        "  thresh = ['Very low (pLDDT < 50)',\n",
        "            'Low (70 > pLDDT > 50)',\n",
        "            'Confident (90 > pLDDT > 70)',\n",
        "            'Very high (pLDDT > 90)']\n",
        "\n",
        "  colors = [x[2] for x in PLDDT_BANDS]\n",
        "\n",
        "  plt.figure(figsize=(2, 2))\n",
        "  for c in colors:\n",
        "    plt.bar(0, 0, color=c)\n",
        "  plt.legend(thresh, frameon=False, loc='center', fontsize=20)\n",
        "  plt.xticks([])\n",
        "  plt.yticks([])\n",
        "  ax = plt.gca()\n",
        "  ax.spines['right'].set_visible(False)\n",
        "  ax.spines['top'].set_visible(False)\n",
        "  ax.spines['left'].set_visible(False)\n",
        "  ax.spines['bottom'].set_visible(False)\n",
        "  plt.title('Model Confidence', fontsize=20, pad=20)\n",
        "  return plt\n",
        "\n",
        "# Show the structure coloured by chain if the multimer model has been used.\n",
        "if model_type_to_use == ModelType.MULTIMER:\n",
        "  multichain_view = py3Dmol.view(width=800, height=600)\n",
        "  multichain_view.addModelsAsFrames(to_visualize_pdb)\n",
        "  multichain_style = {'cartoon': {'colorscheme': 'chain'}}\n",
        "  multichain_view.setStyle({'model': -1}, multichain_style)\n",
        "  multichain_view.zoomTo()\n",
        "  multichain_view.show()\n",
        "\n",
        "# Color the structure by per-residue pLDDT\n",
        "color_map = {i: bands[2] for i, bands in enumerate(PLDDT_BANDS)}\n",
        "view = py3Dmol.view(width=800, height=600)\n",
        "view.addModelsAsFrames(to_visualize_pdb)\n",
        "style = {'cartoon': {'colorscheme': {'prop': 'b', 'map': color_map}}}\n",
        "if show_sidechains:\n",
        "  style['stick'] = {}\n",
        "view.setStyle({'model': -1}, style)\n",
        "view.zoomTo()\n",
        "\n",
        "grid = GridspecLayout(1, 2)\n",
        "out = Output()\n",
        "with out:\n",
        "  view.show()\n",
        "grid[0, 0] = out\n",
        "\n",
        "out = Output()\n",
        "with out:\n",
        "  plot_plddt_legend().show()\n",
        "grid[0, 1] = out\n",
        "\n",
        "display.display(grid)\n",
        "\n",
        "# Display pLDDT and predicted aligned error (if output by the model).\n",
        "if pae_outputs:\n",
        "  num_plots = 2\n",
        "else:\n",
        "  num_plots = 1\n",
        "\n",
        "plt.figure(figsize=[8 * num_plots, 6])\n",
        "plt.subplot(1, num_plots, 1)\n",
        "plt.plot(plddts[best_model_name])\n",
        "plt.title('Predicted LDDT')\n",
        "plt.xlabel('Residue')\n",
        "plt.ylabel('pLDDT')\n",
        "\n",
        "if num_plots == 2:\n",
        "  plt.subplot(1, 2, 2)\n",
        "  pae, max_pae = list(pae_outputs.values())[0]\n",
        "  plt.imshow(pae, vmin=0., vmax=max_pae, cmap='Greens_r')\n",
        "  plt.colorbar(fraction=0.046, pad=0.04)\n",
        "\n",
        "  # Display lines at chain boundaries.\n",
        "  best_unrelaxed_prot = unrelaxed_proteins[best_model_name]\n",
        "  total_num_res = best_unrelaxed_prot.residue_index.shape[-1]\n",
        "  chain_ids = best_unrelaxed_prot.chain_index\n",
        "  for chain_boundary in np.nonzero(chain_ids[:-1] - chain_ids[1:]):\n",
        "    if chain_boundary.size:\n",
        "      plt.plot([0, total_num_res], [chain_boundary, chain_boundary], color='red')\n",
        "      plt.plot([chain_boundary, chain_boundary], [0, total_num_res], color='red')\n",
        "\n",
        "  plt.title('Predicted Aligned Error')\n",
        "  plt.xlabel('Scored residue')\n",
        "  plt.ylabel('Aligned residue')\n",
        "\n",
        "# Save the predicted aligned error (if it exists).\n",
        "pae_output_path = os.path.join(output_dir, 'predicted_aligned_error.json')\n",
        "if pae_outputs:\n",
        "  # Save predicted aligned error in the same format as the AF EMBL DB.\n",
        "  pae_data = notebook_utils.get_pae_json(pae=pae, max_pae=max_pae.item())\n",
        "  with open(pae_output_path, 'w') as f:\n",
        "    f.write(pae_data)\n",
        "\n",
        "# --- Download the predictions ---\n",
        "shutil.make_archive(base_name='prediction', format='zip', root_dir=output_dir)\n",
        "files.download(f'{output_dir}.zip')"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lUQAn5LYC5n4"
      },
      "source": [
        "### Interpretando la predicción\n",
        "\n",
        "En general, el LDDT predicho (pLDDT) se usa mejor para la confianza dentro del dominio, mientras que el error alineado predicho (PAE) se usa mejor para determinar la confianza entre dominios o entre cadenas.\n",
        "\n",
        "Por favor vea el [AlphaFold methods paper](https://www.nature.com/articles/s41586-021-03819-2), el [AlphaFold predictions of the human proteome paper](https://www.nature.com/articles/s41586-021-03828-1), y el [AlphaFold-Multimer paper](https://www.biorxiv.org/content/10.1101/2021.10.04.463034v1) así como [our FAQ](https://alphafold.ebi.ac.uk/faq) sobre cómo interpretar las predicciones de AlphaFold."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "jeb2z8DIA4om"
      },
      "source": [
        "## FAQ & Troubleshooting\n",
        "\n",
        "\n",
        "*   ¿Cómo obtengo una estructura de proteína predicha para mi proteína?\n",
        "    *   Haga clic en el botón _Conectar_ en la parte superior derecha para comenzar.\n",
        "    *   Pegue la secuencia de aminoácidos de su proteína (sin encabezados) en \"Ingrese la secuencia de aminoácidos para plegar\".\n",
        "    *   Ejecute todas las celdas en Colab, ya sea ejecutándolas individualmente (con el botón de reproducción en el lado izquierdo) o a través de _Tiempo de ejecución_ > _Ejecutar todo._ Asegúrese de ejecutar las 5 celdas en orden.\n",
        "    *   La estructura de la proteína predicha se descargará una vez que se hayan ejecutado todas las celdas. Nota: Esto puede llevar de minutos a horas; consulte a continuación.\n",
        "*   ¿Cuánto tiempo llevará esto?\n",
        "    *  La descarga del código fuente de AlphaFold puede tardar unos minutos.\n",
        "    *   La descarga e instalación del software de terceros puede tardar unos minutos.\n",
        "    *   La búsqueda en las bases de datos genéticas puede llevar de minutos a horas.\n",
        "    *   Ejecutar AlphaFold y generar la predicción puede llevar de minutos a horas, dependiendo de la longitud de su proteína y del tipo de GPU que Colab le haya asignado.\n",
        "*   Mi Colab ya no parece estar haciendo nada, ¿qué debo hacer?\n",
        "    *   Algunos pasos pueden tardar de minutos a horas en completarse.\n",
        "    *   Si no sucede nada o si recibe un mensaje de error, intente reiniciar su tiempo de ejecución de Colab a través de _Tiempo de ejecución_ > _Reiniciar tiempo de ejecución_.\n",
        "    *   Si esto no ayuda, intente restablecer su tiempo de ejecución de Colab a través de _Runtime_ > _Factory reset runtime_.\n",
        "*   ¿Cómo se compara esto con la versión de código abierto de AlphaFold?\n",
        "    *   Esta versión de Colab de AlphaFold busca una parte seleccionada del conjunto de datos BFD y actualmente no usa plantillas, por lo que su precisión se reduce en comparación con la versión completa de AlphaFold que se describe en el [AlphaFold paper](https://doi.org/10.1038/s41586-021-03819-2) y [Github repo](https://github.com/deepmind/alphafold/) (la versión completa está disponible a través del script de inferencia).\n",
        "*   ¿Qué es Colab?\n",
        "    *   Ver el [Colab FAQ](https://research.google.com/colaboratory/faq.html).\n",
        "*   Recibí una advertencia \"La computadora portátil requiere mucha RAM\", ¿qué debo hacer?\n",
        "    *  Los recursos asignados a su Colab varían. Ver el [Colab FAQ](https://research.google.com/colaboratory/faq.html) para más detalles.\n",
        "    *   Sin embargo, puedes ejecutar Colab.\n",
        "*   Recibí un error \"El tiempo de ejecución de la CPU de Colab no es compatible\" o \"No se encontró GPU/TPU\", ¿qué debo hacer?\n",
        "    *   El tiempo de ejecución de CPU de Colab no es compatible. Intente cambiar su tiempo de ejecución a través de _runtime> Cambiar tipo de tiempo de ejecución> Acelerador de hardware> GPU.\n",
        "    *  El tipo de GPU asignado a su Colab varía. Ver el [Colab FAQ](https://research.google.com/colaboratory/faq.html) para más detalles.\n",
        "    *  Si recibe \"No se puede conectar al backend de GPU\", puede volver a intentarlo más tarde para ver si Colab le asigna una GPU.\n",
        "    *   [Colab Pro](https://colab.research.google.com/signup) ofrece acceso prioritario a las GPU.\n",
        "*  Recibí un error “ModuleNotFoundError: No module named...”, aunque ejecuté la celda que lo importa, ¿qué hago?\n",
        "    *   Las libretas de Colab en el nivel gratuito caducan después de una cierta cantidad de tiempo. Ver el [Colab FAQ](https://research.google.com/colaboratory/faq.html#idle-timeouts). Intente volver a ejecutar todo el cuaderno desde el principio.\n",
        "*   ¿Esta herramienta instala algo en mi computadora?\n",
        "    *   No, todo sucede en la nube en Google Colab\n",
        "    *   Al final de la ejecución de Colab, se descargará automáticamente en su computadora un archivo zip con la predicción obtenida.\n",
        "*   ¿Cómo debo compartir comentarios e informes de errores?\n",
        "    *   Comparta sus comentarios e informes de errores como un [issue](https://github.com/deepmind/alphafold/issues) en Github.\n",
        "\n",
        "\n",
        "## Trabajo relacionado\n",
        "\n",
        "Eche un vistazo a estos cuadernos de Colab proporcionados por la comunidad (tenga en cuenta que estos cuadernos pueden diferir de nuestro sistema AlphaFold validado y no podemos garantizar su precisión):\n",
        "\n",
        "*   El [ColabFold AlphaFold2 notebook](https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb) por Sergey Ovchinnikov, Milot Mirdita y Martin Steinegger, que utiliza una API alojada en el Södinglab basada en el servidor MMseqs2 ([Mirdita et al. 2019, Bioinformatics](https://academic.oup.com/bioinformatics/article/35/16/2856/5280135)) para la creación de alineación de secuencias múltiples.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "YfPhvYgKC81B"
      },
      "source": [
        "# License and Disclaimer\n",
        "\n",
        "This is not an officially-supported Google product.\n",
        "\n",
        "This Colab notebook and other information provided is for theoretical modelling only, caution should be exercised in its use. It is provided ‘as-is’ without any warranty of any kind, whether expressed or implied. Information is not intended to be a substitute for professional medical advice, diagnosis, or treatment, and does not constitute medical or other professional advice.\n",
        "\n",
        "Copyright 2021 DeepMind Technologies Limited.\n",
        "\n",
        "\n",
        "## AlphaFold Code License\n",
        "\n",
        "Licensed under the Apache License, Version 2.0 (the \"License\"); you may not use this file except in compliance with the License. You may obtain a copy of the License at https://www.apache.org/licenses/LICENSE-2.0.\n",
        "\n",
        "Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.\n",
        "\n",
        "## Model Parameters License\n",
        "\n",
        "The AlphaFold parameters are made available under the terms of the Creative Commons Attribution 4.0 International (CC BY 4.0) license. You can find details at: https://creativecommons.org/licenses/by/4.0/legalcode\n",
        "\n",
        "\n",
        "## Third-party software\n",
        "\n",
        "Use of the third-party software, libraries or code referred to in the [Acknowledgements section](https://github.com/deepmind/alphafold/#acknowledgements) in the AlphaFold README may be governed by separate terms and conditions or license provisions. Your use of the third-party software, libraries or code is subject to any such terms and you should check that you can comply with any applicable restrictions or terms and conditions before use.\n",
        "\n",
        "\n",
        "## Mirrored Databases\n",
        "\n",
        "The following databases have been mirrored by DeepMind, and are available with reference to the following:\n",
        "* UniProt: v2021\\_04 (unmodified), by The UniProt Consortium, available under a [Creative Commons Attribution-NoDerivatives 4.0 International License](http://creativecommons.org/licenses/by-nd/4.0/).\n",
        "* UniRef90: v2022\\_01 (unmodified), by The UniProt Consortium, available under a [Creative Commons Attribution-NoDerivatives 4.0 International License](http://creativecommons.org/licenses/by-nd/4.0/).\n",
        "* MGnify: v2022\\_05 (unmodified), by Mitchell AL et al., available free of all copyright restrictions and made fully and freely available for both non-commercial and commercial use under [CC0 1.0 Universal (CC0 1.0) Public Domain Dedication](https://creativecommons.org/publicdomain/zero/1.0/).\n",
        "* BFD: (modified), by Steinegger M. and Söding J., modified by DeepMind, available under a [Creative Commons Attribution-ShareAlike 4.0 International License](https://creativecommons.org/licenses/by/4.0/). See the Methods section of the [AlphaFold proteome paper](https://www.nature.com/articles/s41586-021-03828-1) for details."
      ]
    }
  ],
  "metadata": {
    "accelerator": "GPU",
    "colab": {
      "private_outputs": true,
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "gpuClass": "standard"
  },
  "nbformat": 4,
  "nbformat_minor": 0
}